redisbp/monitored_client.go

```diff
@@ -0,0 +1,99 @@
+// to support integrating with baseplate.go SpanHooks.
+type MonitoredCmdable interface {
+	redis.Cmdable
+
+	// AddHook adds a hook onto the object.
+	//
+	// Note most redis.Cmdable objects already implement this but it is not a
+	// part of the redis.Cmdable interface.
+	AddHook(hook redis.Hook)
+
+	// WithMonitoredContext returns a clone of the MonitoredCmdable with its
+	// context set to the provided one.
+	//
+	// This is similar to the WithContext method that many redis.Cmdable objects
+	// implement, but this returns a MonitoredCmdable rather than a pointer to
+	// the exact type.  Also note that WithContext is not a part of the
+	// redis.Cmdable interface.
+	WithMonitoredContext(ctx context.Context) MonitoredCmdable
+}
+
+type monitoredClient struct {
+	*redis.Client
+}
+
+func (c *monitoredClient) WithMonitoredContext(ctx context.Context) MonitoredCmdable {
+	return &monitoredClient{Client: c.Client.WithContext(ctx)}
+}
+
+type monitoredCluster struct {
+	*redis.ClusterClient
+}
+
+func (c *monitoredCluster) WithMonitoredContext(ctx context.Context) MonitoredCmdable {
+	return &monitoredCluster{ClusterClient: c.ClusterClient.WithContext(ctx)}
+}
+
+type monitoredRing struct {
+	*redis.Ring
+}
+
+func (c *monitoredRing) WithMonitoredContext(ctx context.Context) MonitoredCmdable {
+	return &monitoredRing{Ring: c.Ring.WithContext(ctx)}
+}
+
+// MonitoredCmdableFactory is used to create Redis clients that are monitored by
+// a SpanHook.
+//
+// A MonitoredCmdableFactory should be created using one of the New methods
+// provided in this package.
+type MonitoredCmdableFactory struct {
+	client MonitoredCmdable
+}
+
+func newMonitoredCmdableFactory(name string, client MonitoredCmdable) MonitoredCmdableFactory {
+	client.AddHook(SpanHook{ClientName: name})
+	return MonitoredCmdableFactory{client: client}
+}
+
+// NewMonitoredClientFactory creates a MonitoredCmdableFactory for a redis.Client
+// object.
+//
+// This may connect to a single redis instance, or be a failover client using
+// Redis Sentinel.
+func NewMonitoredClientFactory(name string, client *redis.Client) MonitoredCmdableFactory {
+	return newMonitoredCmdableFactory(name, &monitoredClient{Client: client})
+}
+
+// NewMonitoredClusterFactory creates a MonitoredCmdableFactory for a
+// redis.ClusterClient object.
+func NewMonitoredClusterFactory(name string, client *redis.ClusterClient) MonitoredCmdableFactory {
+	return newMonitoredCmdableFactory(name, &monitoredCluster{ClusterClient: client})
+}
+
+// NewMonitoredRingFactory creates a MonitoredCmdableFactory for a redis.Ring
+// object.
+func NewMonitoredRingFactory(name string, client *redis.Ring) MonitoredCmdableFactory {
+	return newMonitoredCmdableFactory(name, &monitoredRing{Ring: client})
+}
+
+// BuildClient returns a new MonitoredCmdable with its context set to the
+// provided one.
+func (f MonitoredCmdableFactory) BuildClient(ctx context.Context) MonitoredCmdable {
```

Comment:
This feels like an extra step that's unnecessary. What if we just put this into the 3 constructors (so it's for example `func NewMonitoredClient(ctx context.Context, name string, client *redis.Client) MonitoredCmdable` instead). This way we don't need `MonitoredCmdableFactory` type, and don't need `WithMonitoredContext` inside `MonitoredCmdable` interface.

Reply:
The two things I like about the approach taken here are:

1. You only have to run the "setup" code to add the SpanHook once.
2. (This is the main one) You don't actually have access to the unmonitored, base client so it's harder to accidentally use it.

Reply:
Do you mean `BuildClient` should be called every time a client is used (e.g. it should actually be `factory.BuildClient(ctx).Ping()`, rather than `client := factory.BuildClient(ctx); ... ; client.Ping()`?) If that's the case then yes this is better, but it should be reflected in the example :)

Reply:
It's more targeted at service code where you would do something like:

```go
func (s service) MyEndpoint(ctx context.Context, req controllers.MyEndpointRequest, resp *controllers.MyEndpointResponse) error {
	redis := s.Redis.BuildClient(ctx)
	return controllers.MyEndpoint(ctx, req, resp, redis)
}
```

Where you use it on each new request to build a monitored client to inject into the actual request handler/controller.

Reply:
Ah ok, probably worth explaining that in the doc comment and/or example.

Reply:
Good call, I'll update the examples to reflect the intended usage.